### ISSUE TITLE
unserializable data in a promise resolution can cause a hang

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#4180]
@michaelfig observed a hang, in which one vat never saw the resolution of a Promise from a different vat, despite clear evidence that the Promise was actually resolved. The problem turned out to be that the resolution data was unserializable (it contained an non-`Far`-marked object-with-methods, which throws an error during serialization), and liveslots did not catch/handle the serialization error.

The fix he found was to change the liveslots.js `collect()` function from:
```js
        valueSer = m.serialize(value);
```
to:
```js
      try {
        valueSer = m.serialize(value);
      } catch (e) {
        // Serialization failure.
        valueSer = m.serialize(e);
        rejected = true;
      }
```

This causes liveslots to tell the kernel that the promise was rejected (with a "cannot serialize" Error). This obviously isn't exactly correct (we know for sure the promise was resolved, not rejected), but we can't think of anything better to do. It would be even worse to pretend that it resolved correctly but to some other data. And the serialization error is caught on a turn after the actual `resolve()` was called, so it's far too late to inform userspace somehow.

He built a PR (https://github.com/Agoric/agoric-sdk/pull/4124) with the fix.

cc @erights @michaelfig

**Comment by erights** [comment]
Warning: pedantic distinctions of little practical consequence ahead.

> This obviously isn't exactly correct (we know for sure the promise was resolved, not rejected),

A careful semantics of our distributed object model should be consistent with that, by distinguishing between local and remote promises.

Also, the problem isn't exactly "not serializable", it is "not passable", since it may very well be serializable by marshal parameterized with non-captp-like convertValToSlot and convertSlotToVal functions. The distributed object semantics defines what is a passable value without referring to serialization anyway, and the problem case is exactly when the promise's fulfillment is not passable.

What about the case of a promise rejected with a non-passable reason? This seems similar, but does the above code fix that case?

**Comment by warner** [comment]
> Also, the problem isn't exactly "not serializable", it is "not passable"

In this case, we're watching for an exception from `m.serialize()`, which is the only thing liveslots knows about. So perhaps "not serializable by liveslots" or "not serializable for an eventual send".

> What about the case of a promise rejected with a non-passable reason? This seems similar, but does the above code fix that case?

I've updated the PR's unit test to exercise this case too. You get the same rejection (Error) as for a non-passable non-rejection resolution. (The unit test is nearly identical, it just does `pk1.reject(harden({ nonpassableMethod: () => {} }))` instead of `pk1.resolve(..same)`).

**Comment by warner** [comment]
closed by #4124
